### PR TITLE
Update the eip to store in ring buffer instead of serial storage and add eip 158 handling strategy

### PR DESCRIPTION
**Pull request by g11tech**
* Update the eip to store in ring buffer instead of serial storage 
* add eip 158 handling strategy

### CODE REVIEW

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md]:
```suggestion
        state.insert_slot(HISTORY_STORAGE_ADDRESS, (block.number-1) % HISTORY_SERVE_WINDOW , block.parent.hash)
```

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md:40]:
```suggestion
    if block.timestamp >= FORK_TIMESTAMP and block.number > 0:
```

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md]:
```suggestion
However after weighing pros and cons, we decided to go with just a limited ring buffer to only serve the requiste `HISTORY_SERVE_WINDOW` as [EIP-4788](./eip04788) and beacon state accumulators allow (albit a bit more complex) proof against any ancestor since merge.
```

File: EIPS/eip-2935.md
Comment (gballet) [inline EIPS/eip-2935.md]:
```suggestion
Having contracts (system or otherwise) with hot update paths (branches) poses a risk of "branch" poisioning attacks where one could sprinkle trivial amounts of eth around these hot paths (branches). But it has been deemed that cost of attack would escalate significantly to cause any meaningful slow down of state root updates.
```

**Review by gballet — APPROVED** [review]
Very good. I fixed what I believe are typos. I'll wait for Jochem's feedback before merging.

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:40]: ummm  isn't this a bit redundant? genesis block anyway doesn't undergo processing

File: EIPS/eip-2935.md
Comment (jsign) [inline EIPS/eip-2935.md]: Two notes:
- s/devet/devnet
- I think this should say 5 now.

File: EIPS/eip-2935.md
Comment (jsign) [inline EIPS/eip-2935.md]:
```suggestion
def process_block_hash_history(block: Block, state: State):
```

File: EIPS/eip-2935.md
Comment (jsign) [inline EIPS/eip-2935.md]:
```suggestion
The behavior of `BLOCKHASH` opcode remains same and doesn't affect backward compatibility with the contracts deployed or the gas consumption costs as the resolution of the opcode is handled "directly" by the clients.
```

**Review by jsign — APPROVED** [review]

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md]:
```suggestion
 * for activation at block `1000`, block `744-999`'s hashes will be presisted in the slot and `BLOCKHASH` for `743` or less would resolve to `0` as only `HISTORY_SERVE_WINDOW` can be served.
```

I think the numbers here are incorrect?

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:80]: This is fine, but as note: we will need a similar process like 4788 regarding auditing this contract etc etc in order to ship this :)

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:93]: I'm not sure does this comment apply? Can only get latest 256 blockhashes per this change? I might interpret this comment wrong though?

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [reply EIPS/eip-2935.md:80]: With 4788 there are also checks in the contract (if timestamp does not match storage, then throw). We now seem to treat this contract on a different level: where is the "requested block hash >= block.number - 256" check? Is this in the contract, or is this done "outside" the contract? This is not clear from the current description (at least, for me)

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:102]: I agree and I would honestly think (1) would create an anomaly which we do not want.

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:106]: Is this true? Now `BLOCKHASH` will also invoke some witness costs - right?

**Review by jochem-brouwer — COMMENTED** [review]
Some comments :)

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:80]: contract will implement `resolve_blockhash` logic so yes the check goes inside.

4788 worries about timestamp because could be missed slots, but no such thing as missed numbers (if i understand correctly)

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:93]: yes correct, so we considered serial unbounded storage but only went with ring buffer

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [reply EIPS/eip-2935.md:80]: Yes to be clear: if one would call `resolve_blockhash(n)` and then calldata `n` would just return `sload(n % HISTORY_SERVE_WINDOW)` then we would be in trouble, but if there is this check in the contract that we are within range (i.e. we are within 256 blocks) then we are fine. I think I would just like an explicit note here such that the "shortcut" implementation in clients (i.e. not call the contract but write/read to storage immediately) has a consistent check in the EVM contract :smile:

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:106]: ahh yes, the accesses need to be bundled for stateless clients (will add this) but we can keep the gas same since BLOCKHASH can ideally be resolved/served from maintaining last 256 hashes in memory

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [reply EIPS/eip-2935.md:93]: Ah right sorry, I did not see that this was in the "simplification" part (thus removing the unbounded storage by this PR)

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md:62]: Note: EVM cannot have numbers below 0

File: EIPS/eip-2935.md
Comment (jochem-brouwer) [inline EIPS/eip-2935.md]:
```suggestion
However after weighing pros and cons, we decided to go with just a limited ring buffer to only serve the requisite `HISTORY_SERVE_WINDOW` as [EIP-4788](./eip-4788.md) and beacon state accumulators allow (albeit a bit more complex) proof against any ancestor since merge.
```

File: EIPS/eip-2935.md
Comment (g11tech) [reply EIPS/eip-2935.md:80]: there is a check there, could you confirm if that is correct?

Comment (jochem-brouwer) [reply EIPS/eip-2935.md:80]: If the check is also in the EVM bytecode then it looks correct!

**Review by jochem-brouwer — APPROVED** [review]
LGTM

**Pull request merged**
